source/extensions/filters/http/header_to_metadata/header_to_metadata_filter.h

```diff
@@ -0,0 +1,141 @@
+#pragma once
+
+#include <string>
+#include <tuple>
+#include <vector>
+
+#include "envoy/config/filter/http/header_to_metadata/v2/header_to_metadata.pb.h"
+#include "envoy/server/filter_config.h"
+
+#include "absl/strings/string_view.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace HeaderToMetadataFilter {
+
+enum class MetadataType { String, Number };
```

Comment:
Seems like most of these enums, structs or typedefs are not something you want to publicly expose outside your filter, so they should probably become private members of the Config class. Also, consider just using the Protobuf types for MetadataKeyValue and MetadataType. I don't think the conversion buys you much.

Reply:
I am still undecided about avoiding the conversion from Protobuf
messages to internal structs, mostly because of the LowerCaseString
header field that I'd like to avoid converting/creating on every
request. So there's a small win there — we might also want to extend
things a bit more of which point the conversions might be useful. I
did not make the Rule related structs private because they are shared
between Config and the Filter class.. I could make them private to
Config and make Filter a friend of class but at that point I am
not sure if pays off anymore. Thoughts?

Reply:
I agree that it's worth keeping the LowerCaseString. You could store the protobuf Rule objects directly along side their LowerCaseString names as a `std::vector<std::pair<Http::LowerCaseString, envoy::config::filter::http::header_to_metadata::v2::Config::Rule>>` or with a struct that has the LowerCaseString and the protobuf Rule as fields.

In the future, if there are other pre-computed structures necessary, I think it makes sense to start introducing extra classes.